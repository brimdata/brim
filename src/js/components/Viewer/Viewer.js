/* @flow */

import * as React from "react"

import type {RowRenderer, ViewerDimens} from "../../types"
import Chunk from "./Chunk"
import Chunker from "./Chunker"
import * as Doc from "../../lib/Doc"
import Header from "./Header"
import Log from "../../models/Log"
import ScrollHooks from "../../lib/ScrollHooks"
import * as Styler from "./Styler"
import TableColumns from "../../models/TableColumns"
import {reactElementProps} from "../../test/integration"

type Props = {
  chunker: Chunker,
  dimens: ViewerDimens,
  tableColumns: TableColumns,
  renderRow: RowRenderer,
  logs: Log[],
  onLastChunk?: Function,
  renderEnd: () => *
}

type State = {
  scrollLeft: number,
  chunks: number[]
}

export default class Viewer extends React.PureComponent<Props, State> {
  scrollHooks: Function
  view: ?HTMLDivElement

  constructor(props: Props) {
    super(props)
    this.state = {
      scrollLeft: 0,
      chunks: props.chunker.visibleChunks(0)
    }
    this.scrollHooks = ScrollHooks.create(this.onScrollStart, this.onScrollStop)
  }

  componentDidUpdate() {
    if (!this.view) return
    this.updateChunks(this.view.scrollTop)
    if (
      this.props.chunker.lastChunk() ==
      this.state.chunks[this.state.chunks.length - 1]
    ) {
      this.props.onLastChunk && this.props.onLastChunk()
    }
  }

  onScrollStart() {
    Doc.id("tooltip-root").style.display = "none"
  }

  onScrollStop() {
    Doc.id("tooltip-root").style.display = "block"
  }

  onScroll = () => {
    this.scrollHooks()
    const view = this.view
    if (view) {
      this.updateChunks(view.scrollTop)
      this.setState({scrollLeft: view.scrollLeft})
    }
  }

  updateChunks(scrollTop: number) {
    const next = this.props.chunker.visibleChunks(scrollTop)
    const current = this.state.chunks
    if (!Chunker.isEqual(next, current)) {
      this.setState({chunks: next})
    }
  }

  render() {
    const {
      dimens,
      chunker,
      renderRow,
      logs,
      tableColumns,
      renderEnd
    } = this.props
    const {scrollLeft, chunks} = this.state
    return (
      <div className="viewer" style={{width: dimens.viewWidth}}>
        <Header
          columns={tableColumns}
          dimens={dimens}
          scrollLeft={scrollLeft}
          locator="viewer_header"
        />
        <div
          className="view"
          onScroll={this.onScroll}
          style={{width: dimens.viewWidth, height: dimens.viewHeight}}
          ref={(r) => (this.view = r)}
        >
          <div
            className="list"
            style={Styler.list(dimens)}
            {...reactElementProps("viewer_results")}
          >
            {chunks.map((chunk) => (
              <Chunk
                columns={this.props.tableColumns}
                logs={logs}
                rows={chunker.rows(chunk)}
                key={chunk}
                rowRenderer={renderRow}
                dimens={dimens}
              />
            ))}
            {renderEnd()}
          </div>
        </div>
      </div>
    )
  }
}
