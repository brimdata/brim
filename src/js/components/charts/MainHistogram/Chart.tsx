
import React from "react";

import ChartSVG from "../ChartSVG";
import Dimens from "../../Dimens";
import useMainHistogram from "./useMainHistogram";

export default function MainHistogramChart() {
  return <Dimens className="chart main-search-histogram" render={rect => <MainHistogramSvg width={rect.width} height={rect.height} />} />;
}

const MainHistogramSvg = React.memo(function MainHistogramSvg({
  width,
  height
}) {
  let chart = useMainHistogram(width, height);
  return <ChartSVG chart={chart} />;
});