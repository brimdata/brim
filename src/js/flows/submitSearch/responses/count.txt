{"type":"TaskStart","task_id":0}


{"type":"SearchRecords","channel_id":0,"records":[{"id":46,"type":[{"name":"count","type":"uint64"}],"values":["8100"]}]}


{"type":"SearchEnd","channel_id":0,"reason":"eof"}


{"type":"SearchStats","start_time":{"sec":1598648691,"ns":566035000},"update_time":{"sec":1598648691,"ns":617932000},"bytes_read":4124856,"bytes_matched":3711103,"records_read":11308,"records_matched":8100}


{"type":"TaskEnd","task_id":0}
