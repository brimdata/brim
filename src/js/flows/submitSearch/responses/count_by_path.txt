{"type":"TaskStart","task_id":0}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"type":[{"name":"ts","type":"time"},{"name":"_path","type":"string"},{"name":"count","type":"uint64"}],"values":["1585852140","stats","1"]},{"id":41,"values":["1585852140","dns","15"]},{"id":41,"values":["1585852140","conn","2"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585852080","dns","121"]},{"id":41,"values":["1585852080","conn","58"]},{"id":41,"values":["1585852080","x509","4"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585852020","dns","80"]},{"id":41,"values":["1585852020","conn","12"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851960","conn","16"]},{"id":41,"values":["1585851960","x509","2"]},{"id":41,"values":["1585851960","dns","81"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851900","dns","48"]},{"id":41,"values":["1585851900","conn","17"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851840","dns","38"]},{"id":41,"values":["1585851840","conn","8"]},{"id":41,"values":["1585851840","stats","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851780","dns","42"]},{"id":41,"values":["1585851780","conn","10"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851720","dns","49"]},{"id":41,"values":["1585851720","conn","13"]},{"id":41,"values":["1585851720","x509","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851660","conn","12"]},{"id":41,"values":["1585851660","x509","1"]},{"id":41,"values":["1585851660","dns","71"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851600","conn","23"]},{"id":41,"values":["1585851600","dns","59"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851540","dns","34"]},{"id":41,"values":["1585851540","conn","10"]},{"id":41,"values":["1585851540","stats","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851480","dns","46"]},{"id":41,"values":["1585851480","conn","10"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851420","conn","14"]},{"id":41,"values":["1585851420","dns","86"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851360","dns","149"]},{"id":41,"values":["1585851360","conn","22"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851300","dns","106"]},{"id":41,"values":["1585851300","conn","31"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851240","dns","82"]},{"id":41,"values":["1585851240","conn","22"]},{"id":41,"values":["1585851240","stats","1"]},{"id":41,"values":["1585851240","x509","3"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851180","conn","10"]},{"id":41,"values":["1585851180","dns","71"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851120","conn","15"]},{"id":41,"values":["1585851120","dns","65"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851060","dns","87"]},{"id":41,"values":["1585851060","conn","14"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585851000","dns","64"]},{"id":41,"values":["1585851000","conn","15"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850940","dns","39"]},{"id":41,"values":["1585850940","conn","7"]},{"id":41,"values":["1585850940","stats","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850880","dns","84"]},{"id":41,"values":["1585850880","conn","11"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850820","conn","10"]},{"id":41,"values":["1585850820","dns","70"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850760","dns","93"]},{"id":41,"values":["1585850760","x509","1"]},{"id":41,"values":["1585850760","conn","15"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850700","dns","42"]},{"id":41,"values":["1585850700","conn","13"]},{"id":41,"values":["1585850700","x509","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850640","stats","1"]},{"id":41,"values":["1585850640","dns","74"]},{"id":41,"values":["1585850640","conn","11"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850580","conn","19"]},{"id":41,"values":["1585850580","x509","3"]},{"id":41,"values":["1585850580","dns","87"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850520","conn","10"]},{"id":41,"values":["1585850520","dns","44"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850460","dns","78"]},{"id":41,"values":["1585850460","conn","13"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850400","dns","85"]},{"id":41,"values":["1585850400","conn","14"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850340","dns","79"]},{"id":41,"values":["1585850340","conn","9"]},{"id":41,"values":["1585850340","stats","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850280","x509","2"]},{"id":41,"values":["1585850280","dns","86"]},{"id":41,"values":["1585850280","conn","17"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850220","conn","12"]},{"id":41,"values":["1585850220","dns","54"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850160","dns","176"]},{"id":41,"values":["1585850160","conn","34"]},{"id":41,"values":["1585850160","x509","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850100","dns","69"]},{"id":41,"values":["1585850100","conn","27"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585850040","dns","88"]},{"id":41,"values":["1585850040","conn","49"]},{"id":41,"values":["1585850040","x509","8"]},{"id":41,"values":["1585850040","stats","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849980","conn","5"]},{"id":41,"values":["1585849980","dns","35"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849920","dns","26"]},{"id":41,"values":["1585849920","conn","13"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849860","dns","53"]},{"id":41,"values":["1585849860","conn","13"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849800","dns","41"]},{"id":41,"values":["1585849800","conn","8"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849740","conn","5"]},{"id":41,"values":["1585849740","stats","1"]},{"id":41,"values":["1585849740","dns","35"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849680","dns","39"]},{"id":41,"values":["1585849680","conn","8"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849620","dns","31"]},{"id":41,"values":["1585849620","conn","7"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849560","dns","42"]},{"id":41,"values":["1585849560","conn","13"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849500","conn","8"]},{"id":41,"values":["1585849500","dns","40"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849440","conn","10"]},{"id":41,"values":["1585849440","stats","1"]},{"id":41,"values":["1585849440","dns","42"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849380","conn","11"]},{"id":41,"values":["1585849380","dns","44"]},{"id":41,"values":["1585849380","x509","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849320","conn","10"]},{"id":41,"values":["1585849320","dns","51"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849260","conn","8"]},{"id":41,"values":["1585849260","dns","36"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849200","dns","37"]},{"id":41,"values":["1585849200","conn","5"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849140","stats","1"]},{"id":41,"values":["1585849140","dns","41"]},{"id":41,"values":["1585849140","conn","10"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849080","conn","9"]},{"id":41,"values":["1585849080","dns","41"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585849020","dns","51"]},{"id":41,"values":["1585849020","conn","13"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848960","conn","5"]},{"id":41,"values":["1585848960","dns","33"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848900","conn","15"]},{"id":41,"values":["1585848900","dns","73"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848840","dns","37"]},{"id":41,"values":["1585848840","stats","1"]},{"id":41,"values":["1585848840","conn","6"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848780","dns","48"]},{"id":41,"values":["1585848780","conn","15"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848720","dns","38"]},{"id":41,"values":["1585848720","conn","5"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848660","conn","21"]},{"id":41,"values":["1585848660","dns","104"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848600","conn","10"]},{"id":41,"values":["1585848600","dns","72"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848540","stats","1"]},{"id":41,"values":["1585848540","conn","17"]},{"id":41,"values":["1585848540","dns","110"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848480","dns","74"]},{"id":41,"values":["1585848480","conn","12"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848420","dns","67"]},{"id":41,"values":["1585848420","conn","11"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848360","dns","101"]},{"id":41,"values":["1585848360","conn","10"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848300","dns","86"]},{"id":41,"values":["1585848300","conn","16"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848240","conn","14"]},{"id":41,"values":["1585848240","dns","79"]},{"id":41,"values":["1585848240","stats","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848180","x509","3"]},{"id":41,"values":["1585848180","dns","86"]},{"id":41,"values":["1585848180","conn","21"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848120","x509","1"]},{"id":41,"values":["1585848120","dns","94"]},{"id":41,"values":["1585848120","conn","22"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848060","x509","1"]},{"id":41,"values":["1585848060","dns","107"]},{"id":41,"values":["1585848060","conn","17"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585848000","dns","83"]},{"id":41,"values":["1585848000","conn","16"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585847940","dns","45"]},{"id":41,"values":["1585847940","conn","5"]},{"id":41,"values":["1585847940","stats","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585847880","dns","136"]},{"id":41,"values":["1585847880","conn","13"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585847820","dns","47"]},{"id":41,"values":["1585847820","conn","11"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585847760","dns","136"]},{"id":41,"values":["1585847760","conn","19"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585847700","dns","66"]},{"id":41,"values":["1585847700","conn","42"]},{"id":41,"values":["1585847700","x509","16"]},{"id":41,"values":["1585847700","files","1"]},{"id":41,"values":["1585847700","ssl","1"]},{"id":41,"values":["1585847640","dns","14"]},{"id":41,"values":["1585847640","conn","4"]},{"id":41,"values":["1585847640","stats","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585847580","x509","2"]},{"id":41,"values":["1585847580","dns","50"]},{"id":41,"values":["1585847580","conn","15"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585847520","conn","6"]},{"id":41,"values":["1585847520","dns","26"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585847460","dns","109"]},{"id":41,"values":["1585847460","conn","16"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585847400","conn","21"]},{"id":41,"values":["1585847400","dns","83"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585847340","conn","3"]},{"id":41,"values":["1585847340","dns","71"]},{"id":41,"values":["1585847340","stats","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585847280","dns","110"]},{"id":41,"values":["1585847280","conn","22"]},{"id":41,"values":["1585847220","conn","3"]},{"id":41,"values":["1585847220","dns","7"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585847160","conn","18"]},{"id":41,"values":["1585847160","dns","54"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585847100","dns","54"]},{"id":41,"values":["1585847100","x509","8"]},{"id":41,"values":["1585847100","conn","17"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585847040","x509","1"]},{"id":41,"values":["1585847040","stats","1"]},{"id":41,"values":["1585847040","dns","42"]},{"id":41,"values":["1585847040","conn","13"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585846980","conn","11"]},{"id":41,"values":["1585846980","dns","44"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585846920","x509","1"]},{"id":41,"values":["1585846920","dns","38"]},{"id":41,"values":["1585846920","conn","6"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585846860","dns","46"]},{"id":41,"values":["1585846860","conn","16"]},{"id":41,"values":["1585846860","x509","3"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585846800","dns","37"]},{"id":41,"values":["1585846800","conn","7"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585846740","dns","52"]},{"id":41,"values":["1585846740","conn","15"]},{"id":41,"values":["1585846740","stats","1"]},{"id":41,"values":["1585846680","dns","7"]},{"id":41,"values":["1585846680","conn","7"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585846620","dns","86"]},{"id":41,"values":["1585846620","conn","18"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585846560","dns","42"]},{"id":41,"values":["1585846560","conn","12"]},{"id":41,"values":["1585846560","x509","5"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585846500","dns","114"]},{"id":41,"values":["1585846500","conn","63"]},{"id":41,"values":["1585846500","x509","3"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585846440","dns","53"]},{"id":41,"values":["1585846440","conn","7"]},{"id":41,"values":["1585846440","stats","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585846380","conn","12"]},{"id":41,"values":["1585846380","dns","96"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585846320","dns","48"]},{"id":41,"values":["1585846320","conn","15"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585846260","conn","61"]},{"id":41,"values":["1585846260","dns","99"]},{"id":41,"values":["1585846260","x509","1"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585846200","dns","47"]},{"id":41,"values":["1585846200","conn","15"]}]}


{"type":"SearchRecords","channel_id":0,"records":[{"id":41,"values":["1585846140","conn","4"]},{"id":41,"values":["1585846140","dns","35"]},{"id":41,"values":["1585846140","stats","1"]}]}


{"type":"SearchEnd","channel_id":0,"reason":"eof"}


{"type":"SearchStats","start_time":{"sec":1598648883,"ns":523336000},"update_time":{"sec":1598648883,"ns":577611000},"bytes_read":4124856,"bytes_matched":3711103,"records_read":11308,"records_matched":8100}


{"type":"TaskEnd","task_id":0}
