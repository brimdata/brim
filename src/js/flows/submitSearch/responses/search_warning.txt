{"type":"TaskStart","task_id":0}


{"type":"SearchWarning","warning":"Cut field boo not present in input"}


{"type":"SearchEnd","channel_id":0,"reason":"eof"}


{"type":"SearchEnd","channel_id":0,"reason":"eof"}


{"type":"SearchStats","start_time":{"sec":1598649951,"ns":558286000},"update_time":{"sec":1598649951,"ns":581530000},"bytes_read":4124856,"bytes_matched":4124856,"records_read":11308,"records_matched":11308}


{"type":"TaskEnd","task_id":0}
