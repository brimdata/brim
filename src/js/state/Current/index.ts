import reducer from "./reducer"
import * as selectors from "./selectors"

export default {
  ...selectors,
  reducer
}
