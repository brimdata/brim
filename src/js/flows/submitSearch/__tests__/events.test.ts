import {createZealotMock, zng} from "zealot"

import {submitSearch} from "../mod"
import Chart from "../../../state/Chart"
import Columns from "../../../state/Columns"
import Current from "../../../state/Current"
import Handlers from "../../../state/Handlers"
import SearchBar from "../../../state/SearchBar"
import Spaces from "../../../state/Spaces"
import Viewer from "../../../state/Viewer"
import fixtures from "../../../test/fixtures"
import initTestStore from "../../../test/initTestStore"
import responses from "../../../test/responses"
import Workspaces from "../../../state/Workspaces"

const countByPathResp = responses("count_by_path.txt")
const dnsResp = responses("dns.txt")
const space = fixtures("space1")
const warningResp = responses("search_warning.txt")

let store, zealot, dispatch, select
beforeEach(() => {
  zealot = createZealotMock()
  store = initTestStore(zealot.zealot)
  dispatch = store.dispatch
  select = (s: any) => s(store.getState())

  store.dispatchAll([
    Workspaces.add({
      host: "testHost",
      id: "1",
      name: "testName",
      port: "9867",
      authType: "none"
    }),
    Current.setWorkspaceId("1"),
    Spaces.setDetail("1", space),
    Current.setSpaceId(space.id),
    SearchBar.changeSearchBarInput("dns"),
    SearchBar.pinSearchBar(),
    SearchBar.changeSearchBarInput("query")
  ])
})

const submit = (...args) => dispatch(submitSearch(...args)).catch((e) => e)

describe("table search", () => {
  beforeEach(() => {
    zealot.stubStream("search", countByPathResp).stubStream("search", dnsResp)
  })

  test("zealot gets two requests", async () => {
    await submit()

    const calls = zealot.calls("search")
    expect(calls.length).toBe(2)
    expect(calls[0].args).toEqual("dns query | every 1sec count() by _path")
    expect(calls[1].args).toEqual("dns query | head 500")
  })

  test("the table gets populated", async () => {
    await submit()
    expect(select(Viewer.getViewerRecords).length).toBe(2)
  })

  test("the table gets cleared", async () => {
    dispatch(
      Viewer.setRecords(undefined, [
        new zng.Record([{name: "clear", type: "string"}], ["me"])
      ])
    )
    await submit()
    expect(select(Viewer.getViewerRecords)[0]).not.toEqual([
      {name: "clear", type: "string", value: "me"}
    ])
  })

  test("the table status updates", async () => {
    const promise = submit()
    expect(select(Viewer.getStatus)).toBe("FETCHING")
    expect(select(Viewer.getEndStatus)).toBe("FETCHING")
    await promise
    expect(select(Viewer.getStatus)).toBe("SUCCESS")
    expect(select(Viewer.getEndStatus)).toBe("COMPLETE")
  })

  test("registers a table request then cleans it up", async () => {
    const promise = submit()
    expect(select(Handlers.get)["Table"]).toEqual(
      expect.objectContaining({type: "SEARCH"})
    )
    await promise
    expect(select(Handlers.get)["Table"]).toBe(undefined)
  })

  test("aborts previous table request", async () => {
    const abort = jest.fn()
    dispatch(Handlers.register("Table", {type: "SEARCH", abort}))
    await submit()
    expect(abort).toHaveBeenCalledTimes(1)
  })

  test("sets the viewer columns", async () => {
    await submit()
    expect(select(Viewer.getColumns)).toMatchSnapshot()
  })

  test("sets the Columns columns", async () => {
    await submit()
    expect(select(Columns.getColumns)).toMatchSnapshot()
  })

  test("the chart status updates", async () => {
    const promise = submit()
    expect(select(Chart.getStatus)).toBe("FETCHING")
    await promise
    expect(select(Chart.getStatus)).toBe("SUCCESS")
  })

  test("registers historgram request then cleans it up", async (done) => {
    const promise = submit()
    expect(select(Handlers.get)["Histogram"]).toEqual(
      expect.objectContaining({type: "SEARCH"})
    )
    await promise
    // The promise only waits for the table, might be good to return two
    // promises so people can decide what they want to wait for.
    setTimeout(() => {
      expect(select(Handlers.get)["Histogram"]).toBe(undefined)
      done()
    })
  })

  test("aborts previous histogram request", async () => {
    const abort = jest.fn()
    dispatch(Handlers.register("Histogram", {type: "SEARCH", abort}))
    await submit()
    expect(abort).toHaveBeenCalledTimes(1)
  })

  test("populates the chart", async () => {
    await submit()
    expect(select(Chart.getData)).toMatchSnapshot()
  })
})

describe("search with warnings", () => {
  test("search warnings", async () => {
    zealot
      .stubStream("search", countByPathResp)
      .stubStream("search", warningResp)
    await submit()
    expect(select(SearchBar.getSearchBarError)).toBe(
      "Cut field boo not present in input"
    )
  })
})
