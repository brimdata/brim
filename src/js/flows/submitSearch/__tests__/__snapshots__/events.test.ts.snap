// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table search populates the chart 1`] = `
Object {
  "keys": Array [
    "capture_loss",
    "conn",
    "files",
    "x509",
    "ssl",
    "dns",
    "weird",
    "stats",
  ],
  "table": Object {
    "1582646585000": Object {
      "conn": 1,
      "stats": 1,
    },
    "1582646586000": Object {
      "conn": 1,
      "weird": 1,
    },
    "1582646587000": Object {
      "conn": 2,
    },
    "1582646588000": Object {
      "conn": 5,
    },
    "1582646589000": Object {
      "conn": 1,
    },
    "1582646590000": Object {
      "conn": 2,
    },
    "1582646591000": Object {
      "conn": 1,
      "dns": 1,
    },
    "1582646592000": Object {
      "conn": 1,
    },
    "1582646593000": Object {
      "conn": 2,
      "dns": 1,
    },
    "1582646594000": Object {
      "files": 4,
      "ssl": 1,
      "x509": 1,
    },
    "1582646595000": Object {
      "conn": 3,
    },
    "1582646597000": Object {
      "capture_loss": 1,
    },
  },
}
`;

exports[`table search sets the Columns columns 1`] = `
Object {
  "7247fae6fe9d9e9087c6394e7aba3004": Object {
    "AA:bool": Object {
      "isVisible": true,
    },
    "RA:bool": Object {
      "isVisible": true,
    },
    "RD:bool": Object {
      "isVisible": true,
    },
    "TC:bool": Object {
      "isVisible": true,
    },
    "TTLs:array": Object {
      "isVisible": true,
    },
    "Z:count": Object {
      "isVisible": true,
    },
    "_path:string": Object {
      "isVisible": true,
    },
    "answers:array": Object {
      "isVisible": true,
    },
    "id.orig_h:addr": Object {
      "isVisible": true,
    },
    "id.orig_p:port": Object {
      "isVisible": true,
    },
    "id.resp_h:addr": Object {
      "isVisible": true,
    },
    "id.resp_p:port": Object {
      "isVisible": true,
    },
    "proto:enum": Object {
      "isVisible": true,
    },
    "qclass:count": Object {
      "isVisible": true,
    },
    "qclass_name:string": Object {
      "isVisible": true,
    },
    "qtype:count": Object {
      "isVisible": true,
    },
    "qtype_name:string": Object {
      "isVisible": true,
    },
    "query:string": Object {
      "isVisible": true,
    },
    "rcode:count": Object {
      "isVisible": true,
    },
    "rcode_name:string": Object {
      "isVisible": true,
    },
    "rejected:bool": Object {
      "isVisible": true,
    },
    "rtt:interval": Object {
      "isVisible": true,
    },
    "trans_id:count": Object {
      "isVisible": true,
    },
    "ts:time": Object {
      "isVisible": true,
    },
    "uid:string": Object {
      "isVisible": true,
    },
  },
}
`;

exports[`table search sets the viewer columns 1`] = `
Object {
  "7247fae6fe9d9e9087c6394e7aba3004": Schema {
    "columns": Array [
      Object {
        "name": "_path",
        "type": "string",
      },
      Object {
        "name": "ts",
        "type": "time",
      },
      Object {
        "name": "uid",
        "type": "string",
      },
      Object {
        "name": "id",
        "of": Array [
          Object {
            "name": "orig_h",
            "type": "addr",
          },
          Object {
            "name": "orig_p",
            "type": "port",
          },
          Object {
            "name": "resp_h",
            "type": "addr",
          },
          Object {
            "name": "resp_p",
            "type": "port",
          },
        ],
        "type": "record",
      },
      Object {
        "name": "proto",
        "type": "enum",
      },
      Object {
        "name": "trans_id",
        "type": "count",
      },
      Object {
        "name": "rtt",
        "type": "interval",
      },
      Object {
        "name": "query",
        "type": "string",
      },
      Object {
        "name": "qclass",
        "type": "count",
      },
      Object {
        "name": "qclass_name",
        "type": "string",
      },
      Object {
        "name": "qtype",
        "type": "count",
      },
      Object {
        "name": "qtype_name",
        "type": "string",
      },
      Object {
        "name": "rcode",
        "type": "count",
      },
      Object {
        "name": "rcode_name",
        "type": "string",
      },
      Object {
        "name": "AA",
        "type": "bool",
      },
      Object {
        "name": "TC",
        "type": "bool",
      },
      Object {
        "name": "RD",
        "type": "bool",
      },
      Object {
        "name": "RA",
        "type": "bool",
      },
      Object {
        "name": "Z",
        "type": "count",
      },
      Object {
        "name": "answers",
        "of": "string",
        "type": "array",
      },
      Object {
        "name": "TTLs",
        "of": "interval",
        "type": "array",
      },
      Object {
        "name": "rejected",
        "type": "bool",
      },
    ],
  },
}
`;
